### PR TITLE
Added preserve_names option, fixed position_selector duplication button

### PR DESCRIPTION
**Pull request by maryo**
- Added `preserve_names` option to sort by DOM order while preserving form names
- Fixed `position_selector` duplication button

Symfony normally processes forms in order where it's fields were defined in it's form type so the order of the fields in DOM is not significant. However, there is an open issue with a solution how to handle this case.

https://github.com/symfony/symfony/issues/21595

While i was testing it I encountered a few bugs with combination of `position_selector` option which I have also fixed because the solution is also preserves keys.
I guess it could be a little bit simplified but I did not want to touch your code too much.

Thanks for this repo.

### CODE REVIEW

**Comment by maryo** [comment]
Well. Actually the `findFreeNumericKey` is not needed since `[]` can be used instead. I will update it probably tomorrow.

**Comment by ninsuo** [comment]
Hello maryo,

Thank you for your PR. 

What are the bugs you encountered with `position_selector`? It seems that the new option `preserve_names` you propose does the very same thing.

There were a few bugs in this option, fixed 2 days ago.
Did you have a new try since 2.1.13 release?
Do you reproduce your issues at https://symfony-collection.fuz.org/symfony3/basic/positionField ?

Have a good day.

File: jquery.collection.js [hunk 1]
```diff
@@ -655,6 +673,26 @@
             return elements;
         }
 
+        var getElementKey = function (settings, element) {
+            var name = element.find(':input[name^="' + settings.name_prefix + '["]').attr('name');
```
Comment (ninsuo) [inline jquery.collection.js]: On nested collections, `settings.name_prefix` contains `[` and `]` that will break the jquery expression.

**Comment by ninsuo** [comment]
It seems that the 2.1.13 release was broken though (I pushed the bumping onto another branch...), it may be related. Updated to 2.1.14 just in case. 

Please have a new try.

**Comment by maryo** [comment]
The new option is similar but it supports any key like uuid string keys etc. Not only numeric ones. I was using the new version from master I cloned yesterday. Try the example you linked and duplicate "walk the dog". Then duplicate this new duplicate again. It is inserted at the end instead of third position. And I think it has the same duplicated name but I am on my phone without devtools now. That's what I fixed.

**Comment by ninsuo** [comment]
You are right, thank you for the clarification and the fix 👍 

About "preserve_names", well, it's:

- incompatible with "allow_up" and "allow_down" options if position_field is null because symfony will order using those names

- useless if position_field is not null because it does the same thing

So I propose that we enforce "allow_up" and "allow_down" to false if "preserve_name" is set to true.

**Comment by ninsuo** [comment]
Sounds good, will merge, add the doc and do what I wrote above.
Thanks for your contribution, I've only a few time for this plugin so that's highly appreciated.
Cheers

**Comment by ninsuo** [comment]
About 

> Well. Actually the findFreeNumericKey is not needed since [] can probably be used instead. I will update it probably tomorrow.

I'll let you create a new PR if you feel it needed.

**Comment by maryo** [comment]
It's incompatible and useless by default. But check this issue https://github.com/symfony/symfony/issues/21595
and the RestoreInputOrderExtension. I use it. Reordering then works even without the position field. And perhaps this feature could be solved in future in the Form component.

**Comment by ninsuo** [comment]
I understand the purpose, don't worry.
It just required a bit of improvements to make it cohabit with other options.
I think now it works as expected (currently writing demos on the website).

**Comment by maryo** [comment]
@ninsuo Thank you for such a quick integration 👍 .
But it is compatible with "allow_up" and "allow_down" options as well when the mentioned reordering form type extension is enabled. I was having a limited time yesterday to explain it better. Maybe the options could be set to false by default because it needs the type extension but it should not enforce it so yt can set back to true.
```
// if preserve_names option is set, we should enforce many options to avoid
// having inconsistencies between the UI and the Symfony result
if (settings.preserve_names) {
    settings.allow_up = false;
    settings.allow_down = false;
    settings.drag_drop = false;
    settings.add_at_the_end = true;
 }
```

> Well. Actually the findFreeNumericKey is not needed since [] can probably be used instead. I will update it probably tomorrow.

So actually I realized it is needed :-). `[]` can not be used because
```

**Comment by ninsuo** [comment]
All right, this backend ordering integration makes sense. I may document it as well.

Will remove the options enforcing i added then (but not today weather is great going to enjoy it 😊).

**Pull request merged**